File: internal/store/memory/agent-runtime-state.go
package memory

import (
	"context"
	"time"

	"github.com/rakunlabs/at/internal/service"
)

func (m *Memory) GetAgentRuntimeState(_ context.Context, agentID string) (*service.AgentRuntimeState, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	state, ok := m.agentRuntimeState[agentID]
	if !ok {
		return nil, nil
	}

	return &state, nil
}

func (m *Memory) UpsertAgentRuntimeState(_ context.Context, state service.AgentRuntimeState) error {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	if existing, ok := m.agentRuntimeState[state.AgentID]; ok {
		existing.SessionID = state.SessionID
		existing.StateJSON = state.StateJSON
		existing.LastRunID = state.LastRunID
		existing.LastRunStatus = state.LastRunStatus
		existing.LastError = state.LastError
		existing.UpdatedAt = now
		m.agentRuntimeState[state.AgentID] = existing
	} else {
		state.UpdatedAt = now
		m.agentRuntimeState[state.AgentID] = state
	}

	return nil
}

func (m *Memory) AccumulateUsage(_ context.Context, agentID string, inputTokens, outputTokens, costCents int64) error {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.agentRuntimeState[agentID]
	if !ok {
		existing = service.AgentRuntimeState{
			AgentID: agentID,
		}
	}

	existing.TotalInputTokens += inputTokens
	existing.TotalOutputTokens += outputTokens
	existing.TotalCostCents += costCents
	existing.UpdatedAt = now
	m.agentRuntimeState[agentID] = existing

	return nil
}

File: _ui/src/lib/api/agent-runtime-state.ts
import axios from 'axios';

const api = axios.create({ baseURL: 'api/v1' });

export interface AgentRuntimeState {
  agent_id: string;
  session_id: string;
  state_json: Record<string, any>;
  total_input_tokens: number;
  total_output_tokens: number;
  total_cost_cents: number;
  last_run_id: string;
  last_run_status: string;
  last_error: string;
  updated_at: string;
}

export async function getAgentRuntimeState(agentId: string): Promise<AgentRuntimeState> {
  const res = await api.get<AgentRuntimeState>(`/agents/${agentId}/runtime-state`);
  return res.data;
}

export async function upsertAgentRuntimeState(agentId: string, data: Partial<AgentRuntimeState>): Promise<AgentRuntimeState> {
  const res = await api.put<AgentRuntimeState>(`/agents/${agentId}/runtime-state`, data);
  return res.data;
}

export async function accumulateUsage(agentId: string, data: { input_tokens: number; output_tokens: number; cost_cents: number }): Promise<void> {
  await api.post(`/agents/${agentId}/runtime-state/accumulate`, data);
}
